chain/chain/src/flat_storage_resharder.rs

```diff
@@ -0,0 +1,950 @@
+    pub fn resume(
+        &self,
+        shard_uid: ShardUId,
+        status: &FlatStorageReshardingStatus,
+        scheduler: &dyn FlatStorageResharderScheduler,
+        controller: FlatStorageResharderController,
+    ) -> Result<(), Error> {
+        match status {
+            FlatStorageReshardingStatus::CreatingChild => {
+                // Nothing to do here because the parent will take care of resuming work.
+            }
+            FlatStorageReshardingStatus::SplittingParent(status) => {
+                let parent_shard_uid = shard_uid;
+                info!(target: "resharding", ?parent_shard_uid, ?status, "resuming flat storage shard split");
+                self.check_no_resharding_in_progress()?;
+                // On resume flat storage status is already set.
+                // However, we don't know the current state of children shards,
+                // so it's better to clean them.
+                self.clean_children_shards(&status)?;
+                self.schedule_split_shard(parent_shard_uid, &status, scheduler, controller);
+            }
+            FlatStorageReshardingStatus::CatchingUp(_) => {
+                info!(target: "resharding", ?shard_uid, ?status, "resuming flat storage shard catchup");
+                // TODO(Trisfald): implement child catch up
+                todo!()
+            }
+        }
+        Ok(())
+    }
+
+    /// Starts a resharding event.
+    ///
+    /// For now, only splitting a shard is supported.
+    ///
+    /// # Args:
+    /// * `event_type`: the type of resharding event
+    /// * `shard_layout`: the new shard layout
+    /// * `scheduler`: component used to schedule the background tasks
+    /// * `controller`: manages the execution of the background tasks
+    pub fn start_resharding(
+        &self,
+        event_type: ReshardingEventType,
+        shard_layout: &ShardLayout,
+        scheduler: &dyn FlatStorageResharderScheduler,
+        controller: FlatStorageResharderController,
+    ) -> Result<(), Error> {
+        match event_type {
+            ReshardingEventType::SplitShard(params) => {
+                self.split_shard(params, shard_layout, scheduler, controller)
+            }
+        }
+    }
+
+    /// Starts the event of splitting a parent shard flat storage into two children.
+    fn split_shard(
+        &self,
+        split_params: ReshardingSplitShardParams,
+        shard_layout: &ShardLayout,
+        scheduler: &dyn FlatStorageResharderScheduler,
+        controller: FlatStorageResharderController,
+    ) -> Result<(), Error> {
+        let ReshardingSplitShardParams {
+            parent_shard,
+            left_child_shard,
+            right_child_shard,
+            block_hash,
+            prev_block_hash,
+            ..
+        } = split_params;
+        info!(target: "resharding", ?split_params, "initiating flat storage shard split");
+        self.check_no_resharding_in_progress()?;
```

Comment:
not for this pr
Eventually we'd want to handle the case when there is a at the end of the epoch and resharding is triggered multiple times. In that case you may want to clean up everything and start over.

Reply:
Yeah it could be done through by interrupting through ReshardingHandle or automatically depending on the block hash